File: python/215_kth_largest_element_in_an_array.py
# https://leetcode.com/problems/kth-largest-element-in-an-array/

# Find the kth largest element in an unsorted array. Note that it is the kth largest element in the sorted order, not the kth distinct element.
# For example,
# Given [3,2,1,5,6,4] and k = 2, return 5.
# Note: 
# You may assume k is always valid, 1 ≤ k ≤ array's length.

class Solution(object):
    # An easy AC way. O(n log n).
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        
        nums.sort()
        n = len(nums)
        return nums[n-k]

    # O(n) solution. 
    # Ref: http://www.cs.yale.edu/homes/aspnes/pinewiki/QuickSelect.html
    def findKthLargest_1(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        import random
        pivot = random.choice(nums)
        nums1, nums2 = [], []
        for num in nums:
            if num > pivot:
                nums1.append(num)
            elif num < pivot:
                nums2.append(num)
        if k <= len(nums1):
            return self.findKthLargest(nums1, k)
        if k > len(nums) - len(nums2):
            return self.findKthLargest(nums2, k - (len(nums) - len(nums2)))
        return pivot
        

File: java/215_kth_largest_element_in_an_array.java
public class Solution {
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> queue = new PriorityQueue<Integer>(k);
        for (int num : nums) {
            if (queue.size() < k) {
                queue.add(num);
            } else if (queue.peek()< num) {
                queue.remove();
                queue.add(num);
            }
        }
        return queue.peek();
    }

    public int findKthLargest_1(int[] nums, int k) {
        return findPos(nums, 0, nums.length - 1, k);
    }
    
    private int findPos(int[] nums, int start, int end, int order) {
        int pivot = nums[start];
        int smaller = start;
        int greater = end + 1;
        int index = start + 1;
        while (index < greater) {
            if (nums[index] <= pivot) {
                smaller++;
                index++;
            } else {
                int tmp = nums[--greater];
                nums[greater] = nums[index];
                nums[index] = tmp;
            }
        }
        index--;
        nums[start] = nums[index];
        nums[index] = pivot;
        if (end - index == order - 1) {
            return pivot;
        } else if (end - index < order - 1) {
            return findPos(nums, start, index - 1, order - (end - index) - 1);
        } else {
            return findPos(nums, index + 1, end, order);
        }
    }

    public int findKthLargest_2(int[] nums, int k) {
        Arrays.sort(nums);
        return nums[nums.length-k];
    }
}